In [67]:
import pandas as pd
import numpy as np
from zipfile import ZipFile
from functools import reduce
import matplotlib.pyplot as plt
from scipy import stats
import os

In [68]:
path = r"C:\Users\ncc333\Downloads\archive_Olist.zip"
extracted_file = 'data_files'


In [69]:
with ZipFile(path, "r") as f:
    f.extractall('data_files') #extracts all files in the path provided to a folder names 'data_files'

In [70]:
data_frames = {}

for file in os.listdir(extracted_file):
    if file.endswith('csv'):
        file_path = os.path.join(extracted_file, file) #joins the foldername and each filename into a file path
        data_frames[file] = pd.read_csv(file_path) #this fills the dictionary with filename as the key and the contents of the file as the value

In [71]:
for name, df in data_frames.items():
    print(f'{name}')


olist_customers_dataset.csv
olist_geolocation_dataset.csv
olist_orders_dataset.csv
olist_order_items_dataset.csv
olist_order_payments_dataset.csv
olist_order_reviews_dataset.csv
olist_products_dataset.csv
olist_sellers_dataset.csv
product_category_name_translation.csv


In [72]:
customers = pd.read_csv(r"data_files\olist_customers_dataset.csv") #saves the content of the csv file into the variable named customers
customers.head(2)

# print(customers.columns)
# print(customers.shape)

customer_id                customer_unique_id  \
0  06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1  18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   

   customer_zip_code_prefix          customer_city customer_state  
0                     14409                 franca             SP  
1                      9790  sao bernardo do campo             SP

In [73]:
geolocation = pd.read_csv(r"data_files\olist_geolocation_dataset.csv")
geolocation.head(2)
# print(geolocation.shape)
# print(geolocation.columns)

geolocation_zip_code_prefix  geolocation_lat  geolocation_lng  \
0                         1037       -23.545621       -46.639292   
1                         1046       -23.546081       -46.644820   

  geolocation_city geolocation_state  
0        sao paulo                SP  
1        sao paulo                SP

In [74]:
order_items = pd.read_csv(r"data_files\olist_order_items_dataset.csv")
order_items.head(2)
# print(order_items.shape)
# print(order_items.columns)

order_id  order_item_id  \
0  00010242fe8c5a6d1ba2dd792cb16214              1   
1  00018f77f2f0320c557190d7a144bdd3              1   

                         product_id                         seller_id  \
0  4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
1  e5f2d52b802189ee658865ca93d83a8f  dd7ddc04e1b6c2c614352b383efe2d36   

   shipping_limit_date  price  freight_value  
0  2017-09-19 09:45:35   58.9          13.29  
1  2017-05-03 11:05:13  239.9          19.93

In [75]:
order_payment = pd.read_csv(r"data_files\olist_order_payments_dataset.csv")
order_payment.head(2)
# print(order_payment.columns)
# print(order_payment.shape)

order_id  payment_sequential payment_type  \
0  b81ef226f3fe1789b1e8b2acac839d17                   1  credit_card   
1  a9810da82917af2d9aefd1278f1dcfa0                   1  credit_card   

   payment_installments  payment_value  
0                     8          99.33  
1                     1          24.39

In [76]:
order_reviews = pd.read_csv(r"data_files\olist_order_reviews_dataset.csv")
order_reviews.head(2)
# print(order_reviews.columns)
# print(order_reviews.shape)

review_id                          order_id  \
0  7bc2406110b926393aa56f80a40eba40  73fc7af87114b39712e6da79b0a377eb   
1  80e641a11e56f04c1ad469d5645fdfde  a548910a1c6147796b98fdf73dbeba33   

   review_score review_comment_title review_comment_message  \
0             4                  NaN                    NaN   
1             5                  NaN                    NaN   

  review_creation_date review_answer_timestamp  
0  2018-01-18 00:00:00     2018-01-18 21:46:59  
1  2018-03-10 00:00:00     2018-03-11 03:05:13

In [77]:
order_data = pd.read_csv(r"data_files\olist_orders_dataset.csv")
order_data.head(2)
# print(order_data.columns)
# print(order_data.shape)

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1    delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2018-07-26 14:31:00           2018-08-07 15:27:45   

  order_estimated_delivery_date  
0           2017-10-18 00:00:00  
1           2018-08-13 00:00:00

In [78]:
product = pd.read_csv(r"data_files\olist_products_dataset.csv")
product.head(2)
# print(product.columns)
# print(product.shape)

product_id product_category_name  \
0  1e9e8ef04dbcff4541ed26657ea517e5            perfumaria   
1  3aa071139cb16b67ca9e5dea641aaa2f                 artes   

   product_name_lenght  product_description_lenght  product_photos_qty  \
0                 40.0                       287.0                 1.0   
1                 44.0                       276.0                 1.0   

   product_weight_g  product_length_cm  product_height_cm  product_width_cm  
0             225.0               16.0               10.0              14.0  
1            1000.0               30.0               18.0              20.0

In [79]:
sellers = pd.read_csv(r"data_files\olist_sellers_dataset.csv")
sellers.head(2)
# print(order_reviews.columns)
# print(order_reviews.shape)

seller_id  seller_zip_code_prefix seller_city  \
0  3442f8959a84dea7ee197c632cb2df15                   13023    campinas   
1  d1b65fc7debc3361ea86b5f14c68d2e2                   13844  mogi guacu   

  seller_state  
0           SP  
1           SP

In [80]:
product_category = pd.read_csv(r"data_files\product_category_name_translation.csv")
product_category.head(2)
# print(product_category.columns)
# print(product_category.shape)

product_category_name product_category_name_english
0            beleza_saude                 health_beauty
1  informatica_acessorios         computers_accessories

In [81]:
for name, df in data_frames.items():
    print(f"{name}: {list(df.columns)}") # prints the filename with a list of the columns

olist_customers_dataset.csv: ['customer_id', 'customer_unique_id', 'customer_zip_code_prefix', 'customer_city', 'customer_state']
olist_geolocation_dataset.csv: ['geolocation_zip_code_prefix', 'geolocation_lat', 'geolocation_lng', 'geolocation_city', 'geolocation_state']
olist_orders_dataset.csv: ['order_id', 'customer_id', 'order_status', 'order_purchase_timestamp', 'order_approved_at', 'order_delivered_carrier_date', 'order_delivered_customer_date', 'order_estimated_delivery_date']
olist_order_items_dataset.csv: ['order_id', 'order_item_id', 'product_id', 'seller_id', 'shipping_limit_date', 'price', 'freight_value']
olist_order_payments_dataset.csv: ['order_id', 'payment_sequential', 'payment_type', 'payment_installments', 'payment_value']
olist_order_reviews_dataset.csv: ['review_id', 'order_id', 'review_score', 'review_comment_title', 'review_comment_message', 'review_creation_date', 'review_answer_timestamp']
olist_products_dataset.csv: ['product_id', 'product_category_name', 'pro

In [82]:
#Get sets of columns for each file
columns_per_file = {name: set(df.columns) for name, df in data_frames.items()}
columns_per_file

#Intersection (columns appear in every file)
common_cols = reduce(set.intersection, columns_per_file.values())
print(f"Common columns across all CSVs: {common_cols}")

Common columns across all CSVs: set()


### Merging Datasets

In [83]:
order_data.head(2)

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1    delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2018-07-26 14:31:00           2018-08-07 15:27:45   

  order_estimated_delivery_date  
0           2017-10-18 00:00:00  
1           2018-08-13 00:00:00

In [84]:
customers.head(2)

customer_id                customer_unique_id  \
0  06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1  18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   

   customer_zip_code_prefix          customer_city customer_state  
0                     14409                 franca             SP  
1                      9790  sao bernardo do campo             SP

In [85]:
# Customers + Order
orders_customers = pd.merge(order_data, customers, on="customer_id", how="inner")
orders_customers.head(2)

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1    delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2018-07-26 14:31:00           2018-08-07 15:27:45   

  order_estimated_delivery_date                customer_unique_id  \
0           2017-10-18 00:00:00  7c396fd4830fd04220f754e42b4e5bff   
1           2018-08-13 00:00:00  af07308b275d755c9edb36a90c618231   

   customer_zip_code_prefix customer_city customer_state  
0                      3149     sao paulo             SP  
1                     47813     barreiras             BA

In [86]:
#2 Order + Payments
#this adds payments dataset to the already merged order_customer
orders_payments= pd.merge(orders_customers, order_payment, on="order_id", how="left")
orders_payments.head(2)

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2017-10-04 19:55:00           2017-10-10 21:25:13   

  order_estimated_delivery_date                customer_unique_id  \
0           2017-10-18 00:00:00  7c396fd4830fd04220f754e42b4e5bff   
1           2017-10-18 00:00:00  7c396fd4830fd04220f754e42b4e5bff   

   customer_zip_code_prefix customer_city customer_state  payment_sequential  \
0                      3149     sao paulo             SP                 1.0   
1                      3149     sao paulo             SP                 3.0   

  payment_type  payment_installments  payment_value  
0  credit_card                   1.0          18.12  
1      voucher                   1.0           2.00

In [87]:
#3. Orders + Reviews
orders_reviews = pd.merge(orders_payments, order_reviews, on="order_id", how="left")
orders_reviews.head(2)

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2017-10-04 19:55:00           2017-10-10 21:25:13   

  order_estimated_delivery_date                customer_unique_id  \
0           2017-10-18 00:00:00  7c396fd4830fd04220f754e42b4e5bff   
1           2017-10-18 00:00:00  7c396fd4830fd04220f754e42b4e5bff   

   customer_zip_code_prefix  ... payment_sequential payment_type  \
0                      3149  ...                1.0  credit_card   
1                      3149  ...                3.0      voucher   

   payment_installments payment_value                         review_id  \
0                   1.0         18.12  a54f0611adc9ed256b57ede6b6eb5114   
1                   1.0          2.00  a54f0611adc9ed256b57ede6b6eb5114   

   review_score review_comment_title  \
0           4.0                  NaN   
1           4.0                  NaN   

                              review_comment_message review_creation_date  \
0  Não testei o produto ainda, mas ele veio corre...  2017-10-11 00:00:00   
1  Não testei o produto ainda, mas ele veio corre...  2017-10-11 00:00:00   

  review_answer_timestamp  
0     2017-10-12 03:43:48  
1     2017-10-12 03:43:48  

[2 rows x 22 columns]

In [ ]:
orders_reviews.isna().sum()

order_id                             0
customer_id                          0
order_status                         0
order_purchase_timestamp             0
order_approved_at                  176
order_delivered_carrier_date      1898
order_delivered_customer_date     3154
order_estimated_delivery_date        0
customer_unique_id                   0
customer_zip_code_prefix             0
customer_city                        0
customer_state                       0
payment_sequential                   1
payment_type                         1
payment_installments                 1
payment_value                        1
review_id                          800
review_score                       800
review_comment_title             92482
review_comment_message           61662
review_creation_date               800
review_answer_timestamp            800
dtype: int64

In [88]:
#4. Order + Order Items
orders_items = pd.merge(orders_reviews, order_items, on= "order_id", how="left")
orders_items.head(2)

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2017-10-04 19:55:00           2017-10-10 21:25:13   

  order_estimated_delivery_date                customer_unique_id  \
0           2017-10-18 00:00:00  7c396fd4830fd04220f754e42b4e5bff   
1           2017-10-18 00:00:00  7c396fd4830fd04220f754e42b4e5bff   

   customer_zip_code_prefix  ... review_comment_title  \
0                      3149  ...                  NaN   
1                      3149  ...                  NaN   

                              review_comment_message  review_creation_date  \
0  Não testei o produto ainda, mas ele veio corre...   2017-10-11 00:00:00   
1  Não testei o produto ainda, mas ele veio corre...   2017-10-11 00:00:00   

  review_answer_timestamp  order_item_id                        product_id  \
0     2017-10-12 03:43:48            1.0  87285b34884572647811a353c7ac498a   
1     2017-10-12 03:43:48            1.0  87285b34884572647811a353c7ac498a   

                          seller_id  shipping_limit_date  price freight_value  
0  3504c0cb71d7fa48d967e0e4c94d59d9  2017-10-06 11:07:15  29.99          8.72  
1  3504c0cb71d7fa48d967e0e4c94d59d9  2017-10-06 11:07:15  29.99          8.72  

[2 rows x 28 columns]

In [89]:
orders_items.columns.nunique()

28

In [90]:
# 5. Order Item + Products
orders_products = pd.merge(orders_items, product, on="product_id", how="left")
orders_products.head(2)

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2017-10-04 19:55:00           2017-10-10 21:25:13   

  order_estimated_delivery_date                customer_unique_id  \
0           2017-10-18 00:00:00  7c396fd4830fd04220f754e42b4e5bff   
1           2017-10-18 00:00:00  7c396fd4830fd04220f754e42b4e5bff   

   customer_zip_code_prefix  ...  price freight_value  product_category_name  \
0                      3149  ...  29.99          8.72  utilidades_domesticas   
1                      3149  ...  29.99          8.72  utilidades_domesticas   

  product_name_lenght  product_description_lenght  product_photos_qty  \
0                40.0                       268.0                 4.0   
1                40.0                       268.0                 4.0   

  product_weight_g  product_length_cm product_height_cm product_width_cm  
0            500.0               19.0               8.0             13.0  
1            500.0               19.0               8.0             13.0  

[2 rows x 36 columns]

In [91]:
orders_products.columns.nunique()

36

In [92]:
#6. Products + Product Category Translation
orders_products = pd.merge(
    orders_products, product_category, on="product_category_name",  how="left"
)
orders_products.head(2)


order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2017-10-04 19:55:00           2017-10-10 21:25:13   

  order_estimated_delivery_date                customer_unique_id  \
0           2017-10-18 00:00:00  7c396fd4830fd04220f754e42b4e5bff   
1           2017-10-18 00:00:00  7c396fd4830fd04220f754e42b4e5bff   

   customer_zip_code_prefix  ... freight_value  product_category_name  \
0                      3149  ...          8.72  utilidades_domesticas   
1                      3149  ...          8.72  utilidades_domesticas   

   product_name_lenght product_description_lenght  product_photos_qty  \
0                 40.0                      268.0                 4.0   
1                 40.0                      268.0                 4.0   

   product_weight_g product_length_cm  product_height_cm product_width_cm  \
0             500.0              19.0                8.0             13.0   
1             500.0              19.0                8.0             13.0   

  product_category_name_english  
0                    housewares  
1                    housewares  

[2 rows x 37 columns]

In [93]:
orders_products.columns.nunique()

37

In [96]:
#7. Order Items + Sellers
orders_sellers = pd.merge(orders_products, sellers, on="seller_id", how="left")
orders_sellers.head(2)

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2017-10-04 19:55:00           2017-10-10 21:25:13   

  order_estimated_delivery_date                customer_unique_id  \
0           2017-10-18 00:00:00  7c396fd4830fd04220f754e42b4e5bff   
1           2017-10-18 00:00:00  7c396fd4830fd04220f754e42b4e5bff   

   customer_zip_code_prefix  ... product_description_lenght  \
0                      3149  ...                      268.0   
1                      3149  ...                      268.0   

  product_photos_qty  product_weight_g product_length_cm  product_height_cm  \
0                4.0             500.0              19.0                8.0   
1                4.0             500.0              19.0                8.0   

   product_width_cm product_category_name_english  seller_zip_code_prefix  \
0              13.0                    housewares                  9350.0   
1              13.0                    housewares                  9350.0   

  seller_city seller_state  
0        maua           SP  
1        maua           SP  

[2 rows x 40 columns]

In [100]:
#8. Customers + Geolocation (via zip prefix)
#Note: many geolocations per ZIP, so we take first occurrence per prefix
geo_unique = geolocation.groupby("geolocation_zip_code_prefix").first().reset_index()

final_df = pd.merge(
    orders_sellers, geo_unique, left_on="customer_zip_code_prefix", right_on= "geolocation_zip_code_prefix",
    how="left"
)
print(final_df.shape)
final_df.head()

(119143, 45)


order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
2  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
3  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
4  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
2    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
3    delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
4    delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2017-10-04 19:55:00           2017-10-10 21:25:13   
2          2017-10-04 19:55:00           2017-10-10 21:25:13   
3          2018-07-26 14:31:00           2018-08-07 15:27:45   
4          2018-08-08 13:50:00           2018-08-17 18:06:29   

  order_estimated_delivery_date                customer_unique_id  \
0           2017-10-18 00:00:00  7c396fd4830fd04220f754e42b4e5bff   
1           2017-10-18 00:00:00  7c396fd4830fd04220f754e42b4e5bff   
2           2017-10-18 00:00:00  7c396fd4830fd04220f754e42b4e5bff   
3           2018-08-13 00:00:00  af07308b275d755c9edb36a90c618231   
4           2018-09-04 00:00:00  3a653a41f6f9fc3d2a113cf8398680e8   

   customer_zip_code_prefix  ... product_width_cm  \
0                      3149  ...             13.0   
1                      3149  ...             13.0   
2                      3149  ...             13.0   
3                     47813  ...             19.0   
4                     75265  ...             21.0   

  product_category_name_english  seller_zip_code_prefix     seller_city  \
0                    housewares                  9350.0            maua   
1                    housewares                  9350.0            maua   
2                    housewares                  9350.0            maua   
3                     perfumery                 31570.0  belo horizonte   
4                          auto                 14840.0         guariba   

   seller_state  geolocation_zip_code_prefix geolocation_lat  geolocation_lng  \
0            SP                       3149.0      -23.574809       -46.587471   
1            SP                       3149.0      -23.574809       -46.587471   
2            SP                       3149.0      -23.574809       -46.587471   
3            SP                      47813.0      -12.169860       -44.988369   
4            SP                      75265.0      -16.746337       -48.514624   

  geolocation_city geolocation_state  
0        sao paulo                SP  
1        sao paulo                SP  
2        sao paulo                SP  
3        barreiras                BA  
4       vianopolis                GO  

[5 rows x 45 columns]

## **PROBABILITY**

#### Probability of an Event

Question: What is the probbability that a randomly chosen order recieved a 5-star review

In [109]:
#Probability of a 5-star review
total_reviews = len(orders_reviews)
favourable = (orders_reviews["review_score"]==5).sum()
prob_5star = favourable/total_reviews

print(f"Probability of 5 stars = {round(prob_5star*100,2)}%")

Probability of 5 stars = 57.26%


Interpretation
Approximately 57% of customers give 5 stars

#### Complementary Probability

In [110]:
#What is the probability of not getting a 5-star review?

p_not5 = (orders_reviews["review_score"] !=5).sum()/len(orders_reviews)
print(f"P(Review  )")
print(f"P(Review ≠ 5 stars) = {round(p_not5 *100, 2)}%")

P(Review  )
P(Review ≠ 5 stars) = 42.74%


Approximately 42% of reviews are less than perfect. This highlights that not everyone is fully satisfied wtih product purchased

#### Joint Probability

Concept: P(A∩B)

Question: What is the probability that an order was paid by credit card and received a 5-star review?

In [112]:
orders_reviews.columns

Index(['order_id', 'customer_id', 'order_status', 'order_purchase_timestamp',
       'order_approved_at', 'order_delivered_carrier_date',
       'order_delivered_customer_date', 'order_estimated_delivery_date',
       'customer_unique_id', 'customer_zip_code_prefix', 'customer_city',
       'customer_state', 'payment_sequential', 'payment_type',
       'payment_installments', 'payment_value', 'review_id', 'review_score',
       'review_comment_title', 'review_comment_message',
       'review_creation_date', 'review_answer_timestamp'],
      dtype='object')

In [ ]:
total_orders = len(orders_reviews)
favourable_joint =((orders_reviews['payment_type']=="credit_card") & 
                   (orders_reviews["review_score"]==5)).sum()
p_joint = favourable_joint/total_orders

print(f"P(Credit Card ∩ 5 stars) = {round(p_joint *100,2)}%")


P(Credit Card ∩ 5 stars) = 42.43%


104478

**Interpretation**<br>
Approximately 42% of all orders are paid for by credit card and get a perfect review

#### Conditional Probability

Concept:<br>
`𝑃(𝐴∣𝐵)=𝑃(𝐴∩𝐵)/𝑃(𝐵)` 
                
`𝑃(𝐴∣𝐵)= Favourable_outcomes / Total_credit_card_outcomes`<br>
Question: What is the probability that a review is 5 stars given that the payment was made by credit card?

In [118]:
# If we already know the payment method was credit card, what is the probability that the review is 5 stars?
# This measures satisfaction among credit card users only.

#𝑃(5∣credit_card)
#P(5∣voucher)
#P(5∣boleto)
# Each is a separate conditional distribution. They are not mutually exclusive outcomes of the same experiment.

#this is checking for the probability of getting a 5 star review amongst credit card users, so the total poplation will be the number of credit card users
favourable_cond = ((orders_reviews["payment_type"]=="credit_card") &
                   (orders_reviews["review_score"]==5)).sum()

total_credit = (orders_reviews["payment_type"]=="credit_card").sum()

p_cond = favourable_cond/total_credit
print(f"P(5 stars|Credit Card) = {round(p_cond *100,2)}%")

# this checks the probability of getting a 5 star review amongst voucher users
favourable_cond = ((orders_reviews["payment_type"]=="voucher")&
                   (orders_reviews["review_score"]==5)).sum()
total_credit = (orders_reviews["payment_type"]=="voucher").sum()

p_cond = favourable_cond/total_credit
print(f"P(5 stars|Voucher) = {round(p_cond *100,2)}%")

# this checks the probability of getting a 5 star review amongst boleto users
favourable_cond = ((orders_reviews["payment_type"]=="boleto")&
                   (orders_reviews["review_score"]==5)).sum()
total_credit = (orders_reviews["payment_type"]=="boleto").sum()

p_cond = favourable_cond/total_credit
print(f"P(5 stars|Boleto) = {round(p_cond *100,2)}%")

P(5 stars|Credit Card) = 57.43%
P(5 stars|Voucher) = 55.3%
P(5 stars|Boleto) = 56.95%


**Interpretation**<br>
Each one is around 55-57% meaning in every payment method, about half of the customers gave 5 stars

#### Baye's Rule

Concept:

`P(A∣B )= P(B∣A)⋅P(A) / P(B)`

A = payment

B = review =5

`P(payment | review=5) = P(review=5 | payment).p(payment) / p(review =5)`


Question: Given that an order received a 5-star review, what is the probability that it was paid with each payment method (credit_card, boleto, voucher)?


Among only the 5-star reviews, how are payment methods distributed?

In [120]:
# Among orders that got a 5-star review, what fraction used each payment method (credit_card, boleto, voucher, etc.)?
# That is, if we look at only 5-star reviews,  how are they split across payment types?

#Total number of 5-star reviews (denominator for all)
total_5star = (orders_reviews["review_score"]==5).sum()

# ---- P(payment = credit_card | review = 5) ----
favourable_cond = ((orders_reviews["payment_type"]=="credit_card")&
                   (orders_reviews["review_score"]==5)).sum()
p_cond = favourable_cond/total_5star
print(f"P(Credit Card | 5 stars) = {round(p_cond*100,2)}%")

# ---- P(payment = voucher | review = 5) ----
favourable_cond = ((orders_reviews["payment_type"]=="voucher")&
                   (orders_reviews["review_score"]==5)).sum()
p_cond = favourable_cond/total_5star
print(f"P(Voucher | 5 stars) = {round(p_cond*100,2)}%")

# ---- P(payment = boleto | review = 5) ----
favourable_cond = ((orders_reviews["payment_type"]=="boleto")&
                   (orders_reviews["review_score"]==5)).sum()
p_cond = favourable_cond/total_5star
print(f"P(Boleto | 5 stars) = {round(p_cond*100,2)}%")



P(Credit Card | 5 stars) = 74.11%
P(Voucher | 5 stars) = 5.39%
P(Boleto | 5 stars) = 18.95%


**Interpretation**<br>
Among all 5 star reviews, what fraction used credit card?<br><br>
Credit card - 74.11% of 5 star reviews used credit card<br>
Voucher - 5.39% of 5 star reviews used voucher<br>
Bolero - 18.59% of 5 star reviews used Boleto<br><br>
Unlike P(5 | payment)- conditional probability , these will sum to approximately 100% across all payment methods, because every 5-star order must fall into exactly one payment category.

### Expectation and Variance

**Question1**<br>
What is the average delivery time(expectation) and variance (spread/consistency) of deliveries?

In [121]:
orders_reviews.head()

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
2  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
3  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
4  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
2    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
3    delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
4    delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2017-10-04 19:55:00           2017-10-10 21:25:13   
2          2017-10-04 19:55:00           2017-10-10 21:25:13   
3          2018-07-26 14:31:00           2018-08-07 15:27:45   
4          2018-08-08 13:50:00           2018-08-17 18:06:29   

  order_estimated_delivery_date                customer_unique_id  \
0           2017-10-18 00:00:00  7c396fd4830fd04220f754e42b4e5bff   
1           2017-10-18 00:00:00  7c396fd4830fd04220f754e42b4e5bff   
2           2017-10-18 00:00:00  7c396fd4830fd04220f754e42b4e5bff   
3           2018-08-13 00:00:00  af07308b275d755c9edb36a90c618231   
4           2018-09-04 00:00:00  3a653a41f6f9fc3d2a113cf8398680e8   

   customer_zip_code_prefix  ... payment_sequential payment_type  \
0                      3149  ...                1.0  credit_card   
1                      3149  ...                3.0      voucher   
2                      3149  ...                2.0      voucher   
3                     47813  ...                1.0       boleto   
4                     75265  ...                1.0  credit_card   

   payment_installments payment_value                         review_id  \
0                   1.0         18.12  a54f0611adc9ed256b57ede6b6eb5114   
1                   1.0          2.00  a54f0611adc9ed256b57ede6b6eb5114   
2                   1.0         18.59  a54f0611adc9ed256b57ede6b6eb5114   
3                   1.0        141.46  8d5266042046a06655c8db133d120ba5   
4                   3.0        179.12  e73b67b67587f7644d5bd1a52deb1b01   

   review_score review_comment_title  \
0           4.0                  NaN   
1           4.0                  NaN   
2           4.0                  NaN   
3           4.0     Muito boa a loja   
4           5.0                  NaN   

                              review_comment_message review_creation_date  \
0  Não testei o produto ainda, mas ele veio corre...  2017-10-11 00:00:00   
1  Não testei o produto ainda, mas ele veio corre...  2017-10-11 00:00:00   
2  Não testei o produto ainda, mas ele veio corre...  2017-10-11 00:00:00   
3                               Muito bom o produto.  2018-08-08 00:00:00   
4                                                NaN  2018-08-18 00:00:00   

  review_answer_timestamp  
0     2017-10-12 03:43:48  
1     2017-10-12 03:43:48  
2     2017-10-12 03:43:48  
3     2018-08-08 18:37:50  
4     2018-08-22 19:07:58  

[5 rows x 22 columns]

In [130]:
#Converting the datetime columns to pandas datetime

#Converting to datetime
orders_reviews["order_purchase_timestamp"] = pd.to_datetime(orders_reviews["order_purchase_timestamp"])
orders_reviews["order_delivered_customer_date"] = pd.to_datetime(orders_reviews["order_delivered_customer_date"])

#Calculating delivery time in days
orders_reviews["delivery_time_days"] = (orders_reviews["order_delivered_customer_date"] - orders_reviews["order_purchase_timestamp"]).dt.days

#Dropping missing values
delivery_times = orders_reviews["delivery_time_days"].dropna()

#Getting the expectation
mean_delivery = delivery_times.mean()

#Getting Varinace & standard deviation
variance_delivery = delivery_times.var()
std_delivery = delivery_times.std()

print(f"Expected delivery time (Mean): {round(mean_delivery,2)} days")
print(f"Variance of delivery time: {round(variance_delivery,2)} days")
print(f"Standard Deviation of delivery time: {round(std_delivery,2)} days")

Expected delivery time (Mean): 12.11 days
Variance of delivery time: 91.23 days
Standard Deviation of delivery time: 9.55 days


**Interpretation**<br>
A customer is estimated to recieve their package within approximately 3 days and 22 days, this shows that the delivery company is currently unreliable

**Question 2**<br>
What is the expected customer reviews score and how much do reviews vary?

In [134]:
#Expectation
mean_review = orders_reviews["review_score"].mean()

variance_review =orders_reviews["review_score"].var()
std_review = orders_reviews["review_score"].std()

print(f"Expected review score (Mean): {round(mean_review, 2)}")
print(f"Variance of review score: {round(variance_review, 2)}")
print(f"Standard deviation of review score: {round(variance_review, 2)}")

Expected review score (Mean): 4.08
Variance of review score: 1.82
Standard deviation of review score: 1.82


**Interpretation**<br>
The mean review score show the overall satisfaction (e.g ~4 stars = mostly happy customer). The std tells us that most customers giver between 3 and 5 stars. If Std was very high (e.g 2), it would mean customers are divided (some give 1, others give 5). But the service reputation here is fairly consistent.
